### ISSUE TITLE
CSVBufferIterator potentially used incorrectly

### ISSUE THREAD

**Issue by pitrou** [apache/arrow#14792]
### Describe the bug, including details regarding any error messages, version, and platform.

The CSV streaming reader has this snippet:
https://github.com/apache/arrow/blob/6f4a539323da0d6aef528ccce2404e36f0a08585/cpp/src/arrow/csv/reader.cc#L860-L862

and there's a similar one in the one-shot reader implementation.

This will potentially execute `CSVBufferIterator::operator()` concurrently from multiple threads, but it is not thread-safe as it mutates internal state. Its execution should be serialized.

Fortunately, `CSVBufferIterator` should be very cheap CPU-wise, so instead of transferring it to the CPU executor it can probably be serialized instead from the `IOStream` iterator.

However, there seems to be a similar problem with `SerialBlockReader` which is potentially heavier, especially if `ParseOptions::newlines_in_values` has been enabled.

So perhaps we instead need a more general `SerializingGenerator` building block that delays execution of a generator until its previous execution terminated.

### Component(s)

C++

**Comment by pitrou** [comment]
@westonpace Your insight would be welcome on this.

**Comment by benibus** [comment]
On first glance, it appears that thread-safety would be an issue, but I believe that risk gets removed here:
https://github.com/apache/arrow/blob/6f4a539323da0d6aef528ccce2404e36f0a08585/cpp/src/arrow/csv/reader.cc#L909-L910
`MappingGenerator` seems to have some "hidden" functionality, in that it queues jobs and waits for each each future from its source to finish before proceeding with the next one - so `CSVBufferIterator` and `SerialBlockReader` would never be invoked from multiple threads. I'm not sure if this was an intended side-effect here though.

If I'm not mistaken, this would also imply that `MappingGenerator` is async-reentrant _regardless_ of its source generator - which appears to contradict its documentation.
https://github.com/apache/arrow/blob/6f4a539323da0d6aef528ccce2404e36f0a08585/cpp/src/arrow/util/async_generator.h#L267-L271

**Comment by westonpace** [comment]
@benibus Yes, your understanding matches my own.  I added a similar comment [here](https://github.com/apache/arrow/pull/14269#issuecomment-1343280843).  It does seem the documentation needs to be updated for the mapping generator.

The queuing behavior of mapping generator is nice because it allows a generator to be called async-reentrantly but also not good because it does not forward the async-reentrant pressure, even when you want it.

At one point I had proposed https://issues.apache.org/jira/browse/ARROW-13155 to work around this but I never quite got around to finishing it (and haven't had time to work on the streaming CSV reader since)

**Comment by benibus** [comment]
@westonpace
>The queuing behavior of mapping generator is nice because it allows a generator to be called async-reentrantly

Yeah, it's very useful for that purpose. AFAIK it's the only generator that does this too.

>but also not good because it does not forward the async-reentrant pressure, even when you want it

Right, I guess there wouldn't be an alternative in that case. I suppose that also explains why `BlockParsingOperator` doesn't concern itself with thread-safety either while `BlockDecodingOperator` does (after the initial block) - since it's the "outermost" mapped generator before readahead gets applied.

**Comment by westonpace** [comment]
> Right, I guess there wouldn't be an alternative in that case. I suppose that also explains why BlockParsingOperator doesn't concern itself with thread-safety either while BlockDecodingOperator does (after the initial block) - since it's the "outermost" mapped generator before readahead gets applied.

Even the decoding operator doesn't (I think) need to concern itself with thread-safety.  My original intention was that both decoding and parsing would be thread safe and run in parallel.  However, I never finished that.
